developer_docs/syntactic_stage.md

```diff
@@ -0,0 +1,82 @@
+This function takes several arguments (see docstring for more details).
+The most important arguments are:
+-   _message_ : Describe the issue that lead to the error
+
+-   _symbol_ : The Python ast object should be passed here. This object contains information about its position in the file (line number, column) which ensures the user can more easily locate their error
+
+-   _severity_ : The severity level must be one of the following:
+    -   _warning_ : An error will be printed but Pyccel will continue executing
+    -   _error_ : An error will be printed but Pyccel will continue executing the syntactic stage
+    -   _fatal_ : An error will be printed and Pyccel will stop executing. This level should rarely be needed in the syntactic stage as a failure in one function should not affect the execution of another. It is preferable to show the users all errors at once
+
+## Headers
+
+The headers (type declarations/openmp pragmas/etc) also have their own syntax which cannot be parsed by Python's ast module.
+The module [textx](http://textx.github.io/textX/stable/) is used to parse these statements.
+The files describing the _textx_ grammar are found in the folder [pyccel.parser.grammar](../pyccel/parser/grammar).
+From these files _textx_ generates instances of the classes found in the folder [pyccel.parser.syntax](../pyccel/parser/syntax).
+These instances can then be inserted into the annotated syntax tree.
+
+## Scoping
+
+The final purpose of the syntactic stage is to collect all names used in the code.
+This is important to avoid name collisions if pyccel creates temporaries or requires additional names.
+The names are saved in the scope (for more details see [scope](scope.md)).
+Whenever a symbol is encountered it should be saved to the scope using the function `self.scope.insert_symbol`.
+Any functions visiting a class which inherits from `ScopedNode` must create a new scope before visiting objects and exit it after everything inside the scope has been visited.
+The scope must then be passed to the class using the keyword argument `scope`.
+Care should be taken here as this keyword is not compulsory[^1].
+
+[^1]: The keyword cannot currently be compulsory due to some old code and the special case of `FunctionDef`. A `FunctionDef` which represents a function defined in a header file has no body, and therefore no scope.
+
+A child scope can be created using one of the following functions (for more details see the docstrings in [pyccel.parser.scope](../pyccel/parser/scope.py):
+-   `Scope.new_child_scope`
+-   `Scope.create_new_loop_scope`
+-   `Scope.create_product_loop_scope`
+
+Occasionally it is necessary to create objects in the syntactic stage.
+The `Scope` functions should be used for this purpose to avoid name collisions.
+See the [scope](scope.md) docs for more details.
+In all cases it is preferable to delay this stage as much as possible to ensure that as much information is known about the scope as possible.
```

Comment:
to delay this stage as ~~much~~ **long** as possible

Reply:
Fixed 8bca0416